src/core/config/netdata_publisher.h

```diff
@@ -0,0 +1,155 @@
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ *   This file includes compile-time configurations for Network Data Publisher.
+ *
+ */
+
+#ifndef CONFIG_NETDATA_PUBLISHER_H_
+#define CONFIG_NETDATA_PUBLISHER_H_
+
+#include "config/srp_server.h"
+
+/**
+ * @def OPENTHREAD_CONFIG_NETDATA_PUBLISHER_ENABLE
+ *
+ * Define to 1 to enable Network Data Publisher.
+ *
+ * Network Data Publisher provides mechanisms to limit the number of similar Service and/or Prefix (on-mesh prefix
+ * or external route) entries in the Thread Network Data by monitoring the Network Data and managing if or when to add
+ * or remove entries.
+ *
+ */
+#ifndef OPENTHREAD_CONFIG_NETDATA_PUBLISHER_ENABLE
+#define OPENTHREAD_CONFIG_NETDATA_PUBLISHER_ENABLE OPENTHREAD_CONFIG_SRP_SERVER_ENABLE
+#endif
+
+/**
+ * @def OPENTHREAD_CONFIG_NETDATA_PUBLISHER_MAX_WAIT_TIME_TO_ADD
+ *
+ * Specifies the maximum value (in milliseconds) of the random wait time used before adding an entry in the Thread
+ * Network Data.
+ *
+ */
+#ifndef OPENTHREAD_CONFIG_NETDATA_PUBLISHER_MAX_WAIT_TIME_TO_ADD
+#define OPENTHREAD_CONFIG_NETDATA_PUBLISHER_MAX_WAIT_TIME_TO_ADD 5000
+#endif
+
+/**
+ * @def OPENTHREAD_CONFIG_NETDATA_PUBLISHER_MAX_WAIT_TIME_TO_REMOVE
+ *
+ * Specifies the maximum value (in milliseconds) of the random wait time used before removing an entry from the Thread
+ * Network Data.
+ *
+ */
+#ifndef OPENTHREAD_CONFIG_NETDATA_PUBLISHER_MAX_WAIT_TIME_TO_REMOVE
+#define OPENTHREAD_CONFIG_NETDATA_PUBLISHER_MAX_WAIT_TIME_TO_REMOVE 20000
+#endif
+
+/**
+ * @def OPENTHREAD_CONFIG_NETDATA_PUBLISHER_EXTRA_WAIT_TIME_TO_REMOVE_PREFERRED
+ *
+ * Specifies the extra wait time (in milliseconds) used when removing a preferred entry.
+ *
+ * When removing entries, certain entries are preferred over others. How the preference is determined depends on the
+ * entry type (e.g., (for DNS/SRP service entries associated with smaller RLOC16 or smaller IPv6 address are preferred).
+ * If the `Publisher` determines that its own entry is a preferred one, it adds the above extra wait time in addition
+ * to the random wait selected based on `OPENTHREAD_CONFIG_NETDATA_PUBLISHER_MAX_WAIT_TIME_TO_REMOVE` before  removing
+ * its entry. This gives higher chance for a non-preferred entry from another device to be removed before removing a
```

Comment:
> Can a preferred one never remove itself when the number of entries with same or higher preference is not larger than the threshold? 

This would require/assume that all nodes are running `Publisher` (or similar logic to remove their entry is they see too many similar ones in net data) which may not be the case.